### PR TITLE
event_manager: Add event type flags

### PR DESCRIPTION
**Pull request by zycz**
Added event type flags representing if event type should
be logged, traced and has dynamic data.

Jira: NCSDK-13788

Signed-off-by: Jan Zyczkowski <jan.zyczkowski@nordicsemi.no>

### CODE REVIEW

**Comment by zycz** [comment]
@rakons , @MarekPieta could you please review this PR before I preform changes in all applications. I will have also to describe better the change so that customers will also be able to easily do it.

File: include/event_manager.h
Comment (MarekPieta) [inline include/event_manager.h:37]: Consider using a common prefix for the enum. E.g. `EVENT_TYPE_`

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: include/event_manager.h
Comment (rakons) [reply include/event_manager.h:50]: Yes - this should be removed or at least use directly data from event_type_flags. Now there are magic numbers.

**Review by rakons — CHANGES_REQUESTED** [review]

File: include/event_manager.h
Comment (zycz) [reply include/event_manager.h:37]: > I would also add documentation, because this is part of the user API. Inform that HAS_DYNDATA cannot be set by user too (I would even add build assert to ensure that user does not try to set it).

As said in the beginning I will add documentation and update all our apps after after I will be certain that there will be no more changes.

File: subsys/event_manager/event_manager_priv.h
Comment (zycz) [reply subsys/event_manager/event_manager_priv.h:206]: I cannot use assert here. BUILD_ASSERT would not work because I have to pass to it value from variable.

File: include/event_manager.h
Comment (MarekPieta) [inline include/event_manager.h:48]: Just `Pointer to the event type`

File: include/event_manager.h
Comment (MarekPieta) [inline include/event_manager.h:45]: `Get value of an event type flag`? It returns value of only one flag.

File: include/event_manager.h
Comment (MarekPieta) [inline include/event_manager.h:47]: Maybe just `Selected event type flag` ?

File: include/event_manager.h
Comment (MarekPieta) [reply include/event_manager.h:48]: The event type does not need to be associated with a given event instance that is currently processed.

File: subsys/event_manager/event_manager_priv.h
Comment (MarekPieta) [inline subsys/event_manager/event_manager_priv.h:214]: Maybe we should explicitly clear the flag too (so that we are sure that it always has proper value). Now we do not have a build assert that ensures user did not set this flag.

File: include/event_manager.h
Comment (MarekPieta) [inline include/event_manager.h:40]: Comment is not valid

File: include/event_manager.h
Comment (MarekPieta) [reply include/event_manager.h:45]: Or `Get event type flag's value.`

File: subsys/event_manager/event_manager_priv.h
Comment (zycz) [reply subsys/event_manager/event_manager_priv.h:206]: BUILD_ASSERT((**et_flags**) < (1U << EVENT_FLAGS_SYSTEM_START)

et_flags

**Review by simensrostad — APPROVED** [review]

File: applications/asset_tracker_v2/src/events/app_module_event.c
Comment (pdunaj) [inline applications/asset_tracker_v2/src/events/app_module_event.c:113]: static is needed

Comment (pdunaj) [reply applications/asset_tracker_v2/src/events/app_module_event.c:113]: if we want to keep the variable - check in map if it is optimized out (with limited scope it should be).
I also recommend using short names (c is a spartan language ;)). Try to keep declaration in one line.
e.g.
```
const uint8_t flags = IS_ENABLED(CONFIG_APP_EVENTS_LOG) ? BIT(EVENT_FLAGS_INIT_LOG_ENABLE) : 0;
```

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: include/event_manager.h
Comment (rakons) [reply include/event_manager.h]: But it should be used, if we get rid of the flags definition as a variable.

File: include/event_manager.h
Comment (rakons) [reply include/event_manager.h]: If the documentation is missed - this should be draft. I cannot tick it green if we know that there is something missed.

File: applications/asset_tracker_v2/src/events/app_module_event.c
Comment (zycz) [reply applications/asset_tracker_v2/src/events/app_module_event.c:113]: Yes it is optimized :)

File: applications/nrf_desktop/src/events/passkey_event.c
Comment (zycz) [reply applications/nrf_desktop/src/events/passkey_event.c:61]: I thought that in such way it will be not readable when you will have more flags which you want to send. I also asked you to give such review earlier before I used 2h to manually change EVENT_TYPE_DEFINE (lots of different ways and it was faster to do manually). If you will insist I can change it to the way you suggest.

File: include/event_manager.h
Comment (zycz) [reply include/event_manager.h]: Thanks, for reminding about it. Add change in changelog and forgot to do it in documentation :) 
Added it now.

File: applications/nrf_desktop/src/events/passkey_event.c
Comment (rakons) [reply applications/nrf_desktop/src/events/passkey_event.c]: You can always get rid of the variable and still allow compile time optimisation - use enum value. Look how it was done in my PR: https://github.com/nrfconnect/sdk-nrf/pull/6798/files#diff-116eb9ddb99952b4981b5619fe0f0a31a5493fdeb6301175e4ff269343269866R184. Line 185 in event_manager_priv.h. From the top of my mind the previous version I commented, I was sure the way you are implementing it is there not to change the event definition API. Now as it was changed anyway - it does no more make a big sense to fight with it.

File: doc/nrf/libraries/others/event_manager.rst
Comment (divipillai) [inline doc/nrf/libraries/others/event_manager.rst:147]:
```suggestion
   The :c:macro:`EVENT_TYPE_DEFINE` macro adds flags as a last parameter.
   These flags are constant and can only be set on :c:macro:`EVENT_TYPE_DEFINE` macro.
   To get value of specific flag, use :c:function:`get_event_type_flag` function.
```

**Comment by divipillai** [comment]
Shouldn't the addition of the flag parameter be mentioned in 'Event Manager profiler tracer' as well? 
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/others/event_manager_profiler_tracer.html#implementing-profiling-for-event-manager-events

File: applications/nrf_desktop/src/events/passkey_event.c
Comment (zycz) [reply applications/nrf_desktop/src/events/passkey_event.c]: I have to change TYPE_DEFINE api so that it can easily take some more flags if needed. 

Do I understand correctly that it is ok for you to leave it as it is with flag variable?

File: subsys/event_manager/event_manager.c
Comment (pdunaj) [inline subsys/event_manager/event_manager.c:29]: `BIT(flag)` instead of `(1U << flag)`

File: subsys/event_manager/event_manager_priv.h
Comment (pdunaj) [inline subsys/event_manager/event_manager_priv.h:328]: `s/event_type_flags/flags`

this is a member of `struct event_type` sure this flags are event type flags :)

File: subsys/event_manager/event_manager_priv.h
Comment (pdunaj) [inline subsys/event_manager/event_manager_priv.h:334]: don't add this empty line. the same above with line 329.
try to limit change - simplifies the review

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: applications/asset_tracker_v2/src/events/app_module_event.c
Comment (MarekPieta) [inline applications/asset_tracker_v2/src/events/app_module_event.c:113]:
```
static const uint8_t app_module_event_flags = (IS_ENABLED(CONFIG_APP_EVENTS_LOG)) ?
                                                                           BIT(EVENT_FLAGS_INIT_LOG_ENABLE) : 0;
```
Consider moving `?` to previous line and aligning (for consistency with existing NCS code)

File: applications/asset_tracker_v2/src/events/common_module_event.h
Comment (MarekPieta) [inline applications/asset_tracker_v2/src/events/common_module_event.h:25]: btw. Maybe we could suggest to maintainers that this seem outdated approach which is no longer needed ?

File: doc/nrf/libraries/others/event_manager.rst
Comment (MarekPieta) [inline doc/nrf/libraries/others/event_manager.rst:168]: Define them a few lines above. Otherwise the sample code is incomplete

File: samples/event_manager/src/events/control_event.c
Comment (MarekPieta) [inline samples/event_manager/src/events/control_event.c:20]: Why `__used__`? We can (and even should) opt this out if possible - the value will be stored in the structure that defines event type.

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: samples/event_manager/src/events/control_event.c
Comment (pdunaj) [reply samples/event_manager/src/events/control_event.c:20]: why not static?

File: include/event_manager.h
Comment (pdunaj) [inline include/event_manager.h:44]:
```
	EVENT_TYPE_FLAGS_SYSTEM_START, /* Flags set internally by event manager. */
	EVENT_TYPE_FLAGS_HAS_DYNDATA = EVENT_TYPE_FLAGS_SYSTEM_START,

	EVENT_TYPE_FLAGS_USER_SETTABLE_START, /* Flags available for user. */
	EVENT_TYPE_FLAGS_INIT_LOG_ENABLE = EVENT_TYPE_FLAGS_USER_SETTABLE_START,
	EVENT_TYPE_FLAGS_TRACEABLE,

	EVENT_TYPE_FLAGS_COUNT, /* Number of predefined flags. */

	EVENT_TYPE_FLAGS_USER_DEFINED_START = EVENT_TYPE_FLAGS_COUNT, /* Value greater or equal are available for user. */
```

or

```
	/* Flags set internally by event manager. */
	EVENT_TYPE_FLAGS_SYSTEM_START, 
	EVENT_TYPE_FLAGS_HAS_DYNDATA = EVENT_TYPE_FLAGS_SYSTEM_START,

	/* Flags available for user. */
	EVENT_TYPE_FLAGS_USER_SETTABLE_START,
	EVENT_TYPE_FLAGS_INIT_LOG_ENABLE = EVENT_TYPE_FLAGS_USER_SETTABLE_START,
	EVENT_TYPE_FLAGS_TRACEABLE,

	/* Number of predefined flags. */
	EVENT_TYPE_FLAGS_COUNT,

	/* Value greater or equal are available for user. */
	EVENT_TYPE_FLAGS_USER_DEFINED_START = EVENT_TYPE_FLAGS_COUNT,
```

Comment (pdunaj) [reply include/event_manager.h:44]: remove the comment from line 44, add comments close to each group of flags,
add group that can be used by the user (if we can) - these flags can be useful in hooks

File: applications/asset_tracker_v2/src/events/app_module_event.c
Comment (pdunaj) [inline applications/asset_tracker_v2/src/events/app_module_event.c:112]: I expect we should add flag inside `EVENT_FLAGS_CREATE` - it might be difficult to extend this snippet with new flags
Maybe w can use `IF_ENABLED` macro?

Let's check if following would work:
```
EVENT_FLAGS_CREATE(IF_ENABLED(CONFIG_APP_EVENTS_LOG), EVENT_TYPE_FLAGS_INIT_LOG_ENABLE),
                   IF_ENABLED(CONFIG_SOMETHING), EVENT_TYPE_FLAGS_SOMETHING_ENABLE))
```

File: include/event_manager.h
Comment (pdunaj) [inline include/event_manager.h:59]: let's make it inline static

File: include/event_manager.h
Comment (pdunaj) [inline include/event_manager.h:58]: s/`**/`/`*/`

File: subsys/event_manager/event_manager_priv.h
Comment (pdunaj) [inline subsys/event_manager/event_manager_priv.h:361]: is there a need to move this structure?

Comment (pdunaj) [reply subsys/event_manager/event_manager_priv.h:361]: it's difficult to review :(

File: subsys/event_manager/event_manager.c
Comment (pdunaj) [inline subsys/event_manager/event_manager.c:27]: let's move this code to header file

File: subsys/event_manager/event_manager.c
Comment (pdunaj) [inline subsys/event_manager/event_manager.c:29]: `return et->flags & BIT(flag) != 0;`

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: subsys/event_manager/event_manager_priv.h
Comment (MarekPieta) [inline subsys/event_manager/event_manager_priv.h:299]: Use `BIT_MASK(...)` instead of `BIT(..) - 1`

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: doc/nrf/releases/release-notes-changelog.rst
Comment (zycz) [reply doc/nrf/releases/release-notes-changelog.rst:343]: Discussed offline, I prefer to leave it as it is.

File: doc/nrf/libraries/others/event_manager.rst
Comment (zycz) [reply doc/nrf/libraries/others/event_manager.rst:166]: Discussed offline, use no flags here.

**Review by rakons — APPROVED** [review]

File: applications/asset_tracker_v2/src/events/app_module_event.c
Comment (MarekPieta) [reply applications/asset_tracker_v2/src/events/app_module_event.c:112]: We could do it, but we would need to align Kconfig naming convention for all of the events. In some cases the Kconfig option uses other naming convention with current implementation.

Also keep in mind that using wrong Kconfig option name could be difficult to notice (there is no build error, the event is just e.g. not logged).

File: applications/connectivity_bridge/src/events/uart_data_event.c
Comment (MarekPieta) [inline applications/connectivity_bridge/src/events/uart_data_event.c:28]: Is the indent proper here? Maybe we should use the same convention in the NCS?

File: applications/asset_tracker_v2/src/events/app_module_event.c
Comment (zycz) [reply applications/asset_tracker_v2/src/events/app_module_event.c:112]: It would be also hard to use when other flags will be also set.

File: subsys/event_manager/event_manager_priv.h
Comment (MarekPieta) [inline subsys/event_manager/event_manager_priv.h:307]: CONFIG_EVENT_MANAGER_TRACE_EVENT_DATA should still be used here.

File: tests/subsys/event_manager/src/events/sized_events.c
Comment (MarekPieta) [inline tests/subsys/event_manager/src/events/sized_events.c:2]: Fix, please :)

**Review by MarekPieta — APPROVED** [review]
A few last fixes. Please apply them.

File: applications/connectivity_bridge/src/events/ble_ctrl_event.c
Comment (MarekPieta) [inline applications/connectivity_bridge/src/events/ble_ctrl_event.c:24]: Still not aligned.

File: include/event_manager.h
Comment (MarekPieta) [inline include/event_manager.h:51]: Discussed offline. Consider removing it, we can add it later if needed.

**Review by MarekPieta — APPROVED** [review]

File: west.yml
Comment (pdunaj) [inline west.yml:149]: remember to change it before merging

**Review by pdunaj — APPROVED** [review]

File: doc/nrf/libraries/others/event_manager.rst
Comment (divipillai) [inline doc/nrf/libraries/others/event_manager.rst:147]:
```suggestion
   If you do not want to set any flag, use :c:macro:`EVENT_FLAGS_CREATE` without any argument as shown in the below example.
```
Just minor updates

File: doc/nrf/releases/release-notes-changelog.rst
Comment (divipillai) [inline doc/nrf/releases/release-notes-changelog.rst:354]:
```suggestion
    * Event type flags to represent if event type should be logged, traced and has dynamic data.
      To update your application, pass a flag variable as a parameter in :c:macro:`EVENT_TYPE_DEFINE` instead of ``init_log``.
      You can use :c:macro:`EVENT_FLAGS_CREATE` to set multiple flags:
```

File: doc/nrf/releases/release-notes-changelog.rst
Comment (divipillai) [inline doc/nrf/releases/release-notes-changelog.rst:372]: Shouldn't there be a change log entry for event manager profiler tracer as well?

File: doc/nrf/libraries/others/event_manager.rst
Comment (zycz) [reply doc/nrf/libraries/others/event_manager.rst:147]: I think it should be: "To **not** set any flag...."

File: doc/nrf/releases/release-notes-changelog.rst
Comment (zycz) [reply doc/nrf/releases/release-notes-changelog.rst:372]: Rather not. We change defining flags in many applications and samples yet it is a change related to event_manager API. event_manager_profiler_tracer only uses this API.

**Review by divipillai — APPROVED** [review]

**Pull request merged**
